pkg/vm/compiler/codegen.go

```diff
@@ -382,15 +382,24 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 			ast.Walk(c, n.X)
 			ast.Walk(c, n.Y)
 
-			// VM has separate opcode for string concatenation
-			if n.Op == token.ADD {
+			switch {
+			case n.Op == token.ADD:
+				// VM has separate opcodes for number and string concatenation
 				typ, ok := tinfo.Type.Underlying().(*types.Basic)
 				if ok && typ.Kind() == types.String {
 					emitOpcode(c.prog, vm.CAT)
 				} else {
 					emitOpcode(c.prog, vm.ADD)
 				}
-			} else {
+			case n.Op == token.EQL:
+				// VM has separate opcodes for number and string equality
+				typ, ok := tinfo.Type.Underlying().(*types.Basic)
+				if ok && typ.Kind() == types.String {
```

Comment:
Makes me wonder about other types like byte arrays, are they gonna be treated properly by this code?

Reply:
And BTW, we have `isStringType()` that probably can simply code a bit here.

Reply:
Yes, with `isStringType` it looks nicer.
Regarding your first point, in `Go` slices cannot be compared with `==` (except with `nil`).
For the nil case, I haven't yet checked how will it work.